Mathlib/NumberTheory/DirichletCharacter/Basic.lean

```diff
@@ -0,0 +1,51 @@
+/-
+Copyright (c) 2023 Ashvni Narayanan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ashvni Narayanan, Moritz Firsching
+-/
+import Mathlib.NumberTheory.LegendreSymbol.MulCharacter
+import Mathlib.Data.ZMod.Algebra
+
+/-!
+# Dirichlet Characters
+
+Let `R` be a commutative monoid with zero. A Dirichlet character `χ` of level `n` over `R` is a
+multiplicative character from `ZMod n` to `R` sending non-units to 0. We then obtain some properties
+of `toUnitHom χ`, the restriction of `χ` to a group homomorphism `(ZMod n)ˣ →* Rˣ`.
+
+Main definitions:
+
+- `DirichletCharacter`: The type representing a Dirichlet character.
+
+## TODO
+
+- `change_level`: Extend the Dirichlet character χ of level `n` to level `m`, where `n` divides `m`.
+- definition of conductor
+
+## Tags
+
+dirichlet character, multiplicative character
+-/
+
+/-- The type of Dirichlet characters of level `n`. -/
+abbrev DirichletCharacter (R : Type) [CommMonoidWithZero R] (n : ℕ) := MulChar (ZMod n) R
+
+open MulChar
+variable {R : Type} [CommMonoidWithZero R] {n : ℕ} (χ : DirichletCharacter R n)
+
+namespace DirichletCharacter
+lemma toUnitHom_eq_char' {a : ZMod n} (ha : IsUnit a) :
+  χ a = χ.toUnitHom ha.unit := by simp
+
+lemma toUnitHom_eq_iff (ψ : DirichletCharacter R n) :
+  toUnitHom χ = toUnitHom ψ ↔ χ = ψ := by simp
+
+lemma eval_modulus_sub (x : ZMod n) :
+  χ (n - x) = χ (-x) := by simp
+
+lemma isPeriodic {m : ℕ} (hm : n ∣ m) (a : ℤ) :
```

Comment:
Maybe we want to use [Function.Periodic](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Periodic.html#Function.Periodic) (to have all the API)?

Reply:
done: now using `Function.Periodic`

Reply:
Looking at stuff like [Real.sin_periodic](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/SpecialFunctions/Trigonometric/Basic.html#Real.sin_periodic) I don't see a clear convention.. maybe just `periodic`, but I don't think it is very important, we can always change it later.

Reply:
ok, changing it to `periodic`